# Practice Exercises

In [1]:
# Import Libraries 
import pandas as pd 
import numpy as np 
import os 

In [2]:
# Create path 
path = r'C:\Users\Megan\Documents\CareerFoundry\Unit 4\InstaCart Python Project'

In [3]:
# Import products.csv 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import orders_wrangled.csv 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week    order_hour  days_since_prior_order  
count       3.421083e+06  3.421083e+06            3.214874e+06  
mean        2.776219e+00  1.345202e+01            1.111484e+01  
std         2.046829e+00  4.226088e+00            9.206737e+00  
min         0.000000e+00  0.000000e+00            0.000000e+00  
25%         1.000000e+00  1.000000e+01            4.000000e+00  
50%         3.000000e+00  1.300000e+01            7.000000e+00  
75%         5.000000e+00  1.600000e+01            1.500000e+01  
max         6.000000e+00  2.300000e+01            3.000000e+01

In [6]:
# Create a dataframe 
df_test = pd.DataFrame()

In [7]:
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check column for mixed-type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Change data type of mix column to string
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
df_test.dtypes

mix    object
dtype: object

In [12]:
# Count missing values 
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Create subset with missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape


(49693, 5)

In [16]:
# Create subset without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Find duplicates 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
# Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
# Export checked products.csv 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Products_Checked.csv'))

# Task 4.5


In [24]:
# 01. Check for inconsistencies in df_ords 
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week    order_hour  days_since_prior_order  
count       3.421083e+06  3.421083e+06            3.214874e+06  
mean        2.776219e+00  1.345202e+01            1.111484e+01  
std         2.046829e+00  4.226088e+00            9.206737e+00  
min         0.000000e+00  0.000000e+00            0.000000e+00  
25%         1.000000e+00  1.000000e+01            4.000000e+00  
50%         3.000000e+00  1.300000e+01            7.000000e+00  
75%         5.000000e+00  1.600000e+01            1.500000e+01  
max         6.000000e+00  2.300000e+01            3.000000e+01

In [25]:
df_ords.head(100-110)

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421068     3421068    803273   206208            50                  5   
3421069     3421069   3154581   206209             1                  3   
3421070     3421070   1889163   206209             2                  3   
3421071     3421071   1542354   206209             3                  5   
3421072     3421072    688306   206209             4                  1   

         order_hour  days_since_prior_order  
0                 8                     NaN  
1                 7                    15.0  
2                12                    21.0  
3                 7                    29.0  
4                15                    28.0  
...             ...                     ...  
3421068          11                     4.0  
3421069          11                     NaN  
3421070          17                     7.0  
3421071          11                    30.0  
3421072          10                    30.0  

[3421073 rows x 7 columns]

## 02. Although there are more than 3,000,000 records in the dataframe, the order numbers continually repeat between 1-100. This is odd considering that each order should have a unique order number 

In [26]:
# 03. Check for mixed data in df_ords 
for col in df_ords.columns.tolist(): 
    weird = (df_ords[[col]].applymap(type) !=df_ords
            [[col]].iloc[0].apply(type)).any(axis = 1) 
    if len(df_ords[weird]) > 0: print(col)

## 03. There are no mixed data columns in df_ords

In [27]:
# 05. Look for missing values in df_ords 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour                     0
days_since_prior_order    206209
dtype: int64

## 05. There are 206,209 missing values in the column days_since_prior_order. These values could be missing because they are from new customers who have no prior orders. 


In [28]:
# 06. Impute missing values using the mean of the column, 11 days 
df_ords['days_since_prior_order'].fillna(11, inplace = True)

In [29]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  order_hour  \
0           0   2539329        1             1                  2           8   
1           1   2398795        1             2                  3           7   
2           2    473747        1             3                  3          12   
3           3   2254736        1             4                  4           7   
4           4    431534        1             5                  4          15   

   days_since_prior_order  
0                    11.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [30]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week    order_hour  days_since_prior_order  
count       3.421083e+06  3.421083e+06            3.421083e+06  
mean        2.776219e+00  1.345202e+01            1.110791e+01  
std         2.046829e+00  4.226088e+00            8.924994e+00  
min         0.000000e+00  0.000000e+00            0.000000e+00  
25%         1.000000e+00  1.000000e+01            5.000000e+00  
50%         3.000000e+00  1.300000e+01            8.000000e+00  
75%         5.000000e+00  1.600000e+01            1.500000e+01  
max         6.000000e+00  2.300000e+01            3.000000e+01

## 06. I chose to use the mean to impute the missing values because it won't dramatically change the overall descriptive statistics of the column. The only other option would have been to drop the records with missing values, but that would have impacted the descriptive statistics of all the other columns. 


In [31]:
# 07. Look for duplicates 
df_dups = df_ords[df_ords.duplicated()]

In [32]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour, days_since_prior_order]
Index: []

## 07. There are no duplicate rows in the dataframe df_ords. 


## 08. There are no duplicate rows to address

In [33]:
df_ords.shape

(3421083, 7)

In [34]:
# 09. Export df_ords 
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

## 09. df_prods_clean_no_dups already exported in the 'Exercises' section